layer/csv_logging.h

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef STADIA_OPEN_SOURCE_PERFORMANCE_LAYERS_CSV_LOGGING_H_
+#define STADIA_OPEN_SOURCE_PERFORMANCE_LAYERS_CSV_LOGGING_H_
+
+#include <cstdio>
+#include <string>
+
+#include "event_logging.h"
+#include "layer_data.h"
+
+namespace performancelayers {
+// Takes an `Event` instance as an input and generates a csv string containing
+// `event`'s name and attribute values.
+// TODO(miladhakimi): Differentiate hashes and other integers. Hashes
+// should be displayed in hex.
+std::string EventToCSVString(Event &event);
+
+// CSVLogger logs the events in the CSV format to the output given in its
+// constructor. Sample use:
+// ```c++
+// CSVLogger logger("pipeline,duration", "compile_time.csv");
+// logger.StartLog();
+// Event compile_time_event = ...;
+// logger.AddEvent(&compile_time_event);
+// logger.Flush();
+// logger.EndLog();
+// ```
+// There is no need to add '\n' at the end of the csv_header in the constructor.
+// This is handled by the implementation.
+class CSVLogger : public EventLogger {
+ public:
+  CSVLogger(const char *csv_header, const char *filename)
+      : header_(csv_header) {
+    if (filename) {
+      out_ = fopen(filename, "w");
+    } else {
+      out_ = stderr;
+    }
+  }
+
+  ~CSVLogger() {
+    if (out_ != stderr) fclose(out_);
```

Comment:
Shouldn't we close the file on `EndLog` and reset the handle? Here we could conditionally close it if `OnEnd` wasn't called

Reply:
Since we're opening it in the constructor and once, we can only close it on the destructor. If I have it on `EndLog()`, I also need to check if the file is closed  in the destructor and close it if it isn't (the `EndLog()` was not called). I also don't need to worry about multiple `EndLog()` invocations.

Reply:
> Since we're opening it in the constructor and once, we can only close it on the destructor.
Why? This is what `std::fstream` does on `.close()`: https://en.cppreference.com/w/cpp/io/basic_fstream

> I also need to check if the file is closed in the destructor and close it if it isn't
We can set the file handle to nullptr and close only when not null and not stderr.

Reply:
> I also don't need to worry about multiple EndLog() invocations.

I see your point but I think it would be easier to just make `EndLog` support this scenario. From the practical perspective it may be convenient to make double EndLog a noop and only disallow adding event after end

Reply:
> > Since we're opening it in the constructor and once, we can only close it on the destructor.
> 
> Why? This is what `std::fstream` does on `.close()`: https://en.cppreference.com/w/cpp/io/basic_fstream
> 
> > I also need to check if the file is closed in the destructor and close it if it isn't
> 
> We can set the file handle to nullptr and close only when not null and not stderr.

Cool! I didn't know it closes the file in the destructor.

Reply:
> We can set the file handle to nullptr and close only when not null and not stderr.

Yep, That's what I did first then I changed it to this. I'll fix it.